src/ui/public/indexed_array/indexed_array.js

```diff
@@ -1,156 +1,165 @@
 
 /**
- * Generic extension of Array class, which will index (and reindex) the
- * objects it contains based on their properties.
+ *  Generic extension of Array class, which will index (and reindex) the
+ *  objects it contains based on their properties.
  *
- * @class  IndexedArray
- * @module utils
- * @constructor
- * @param {object}   [config]            - describes the properties of this registry object
- * @param {string[]} [config.index]      - a list of props/paths that should be used to index the docs.
- * @param {string[]} [config.group]      - a list of keys/paths to group docs by.
- * @param {string[]} [config.order]      - a list of keys/paths to order the keys by.
- * @param {object[]} [config.initialSet] - the initial dataset the IndexedArray should contain.
- * @param {boolean}  [config.immutable]  - a flag that hints to people reading the implementation
- *                                       that this IndexedArray should not be modified. It's modification
- *                                       methods are also removed
+ *  @param {Object} config describes the properties of this registry object
+ *  @param {Array<string>} [config.index] a list of props/paths that should be used to index the docs.
+ *  @param {Array<string>} [config.group] a list of keys/paths to group docs by.
+ *  @param {Array<string>} [config.order] a list of keys/paths to order the keys by.
+ *  @param {Array<any>} [config.initialSet] the initial dataset the IndexedArray should contain.
+ *  @param {boolean} [config.immutable] a flag that hints to people reading the implementation that this IndexedArray
+ *                                      should not be modified
  */
-_.class(IndexedArray).inherits(Array);
-export function IndexedArray(config) {
-  IndexedArray.Super.call(this);
+export class IndexedArray extends Array {
+  static OPT_NAMES = OPT_NAMES
 
-  // just to remind future us that this list is important
-  config = _.pick(config || {}, OPT_NAMES);
+  constructor(config) {
+    super();
 
-  this.raw = [];
+    config = _.pick(config || {}, OPT_NAMES);
 
-  // setup indices
-  this._indexNames = _.union(
-    this._setupIndices(config.group, inflectIndex, _.organizeBy),
-    this._setupIndices(config.index, inflectIndex, _.indexBy),
-    this._setupIndices(config.order, inflectOrder, _.sortBy)
-  );
+    // use defineProperty so that value can't be changed
+    Object.defineProperty(this, 'raw', { value: [] });
 
-  if (config.initialSet) {
-    this.push.apply(this, config.initialSet);
-  }
+    this._indexNames = _.union(
+      this._setupIndex(config.group, inflectIndex, _.organizeBy),
+      this._setupIndex(config.index, inflectIndex, _.indexBy),
+      this._setupIndex(config.order, inflectOrder, _.sortBy)
+    );
 
-  if (config.immutable) {
-    // just a hint, bugs caused by updates not propogating would be very
-    // very very hard to track down
-    this.push = this.splice = undefined;
```

Comment:
Yeah, they are now implemented as methods and throw descriptive errors when immutable is turned on (it also covers all of the mutable methods, not just push and splice).